OpenTerm/Controller/ViewController.swift

```diff
@@ -339,6 +357,261 @@ extension ViewController: UIDocumentPickerDelegate {
+            
+            alertController.addAction(cancelAction)
+            
+            sender.present(alertController,
+                           animated: true,
+                           completion: nil)
+        }
+    }
+    
+    /// Saves the passed url as a bookmark.
+    ///
+    /// - Parameter url: The url to be saved as a bookmark.
+    func saveBookmarkURL(url: URL) throws {
+        /**
+         *  Getting the bookmark data for the current URL can fail. E.g.,
+         *  when we don't have access to the corresponding security scoped
+         *  resource. However, since the user can only save a directory URL
+         *  that is currently accessed, this should not happen.
+         */
+        do {
+            /**
+             *  Get the bookmark data for the URL making sure it is suitable
+             *  to be saved as a file.
+             */
+            let bookmark = try url.bookmarkData(options: .suitableForBookmarkFile,
+                                                includingResourceValuesForKeys: nil,
+                                                relativeTo: nil)
+            
+            /**
+             *  Get the document directory of the app. The bookmarks will be saved
+             *  in a hidden folder there.
+             */
+            if let dir = FileManager.default.urls(for: .documentDirectory,
+                                                  in: .userDomainMask).first {
+                
+                /**
+                 *  Get the URL of the bookmark directory (where the bookmarks are saved).
+                 */
+                let bookmarkDirectoryURL = dir.appendingPathComponent(ViewController.bookmarkDirectory,
+                                                                      isDirectory: true)
+                
+                /**
+                 *  Create the bookmarks directory (if it doesn't exist)
+                 */
+                try FileManager.default.createDirectory(at: bookmarkDirectoryURL,
+                                                        withIntermediateDirectories: true,
+                                                        attributes: nil)
+                
+                /**
+                 *  The URL for where the bookmark data will be saved.
+                 */
+                let bookmarkDataURL = bookmarkDirectoryURL.appendingPathComponent(url.lastPathComponent,
+                                                                                  isDirectory: false)
+                
+                /**
+                 *  Actually saving the bookmark data.
+                 */
+                try URL.writeBookmarkData(bookmark, to: bookmarkDataURL)
+                
+                /**
+                 *  Tell the bookmark view controller that the bookmarks were updated.
+                 */
+                self.bookmarkViewController.bookmarksWereUpdated()
+            }
+        } catch {
+            throw error
```

Comment:
If you just re-throw the error, why bother with `do { ... } catch { }` at all?

Reply:
Oh, good _catch_.^^ Didn't know it would just pass the throw on. Will change that.